rcos/Telescope
# 128 add generate command

Issue Number: close #128 

Problem Summary:
Add `/generate` slash command to Discord bot.

### What is changed and how it works?
**graphql/rcos/discord_assoications, src/discord_bot/commands, src/api/rcos/discord_assoications: "generate.rs",...**
#### generate.rs:
- Implemented `/generate` command.
- `/generate` command has 4 subcommands: `/generate roles`, `/generate categories`, `/generate channels` and `/generate all`:
  - ##### `/generate roles`:
    - Creates roles for projects and/or small groups in the ongoing semester.
  - ##### `/generate categories`:
    - Creates categories for projects and/or small groups in the ongoing semester. 
    - Categories are private and are only accessible by project/small group member, Faculty Advisors, Coordinators and    Sysadmin.
  - ##### `/generate channels`:
    - creates voice and text channels for projects and/or small groups in the ongoing semester. 
    - If category for project/small group is created, channels will be generated under the project/small group category.
    - Channels are private and are only accessible by project/small group member, Faculty Advisors, Coordinators and    Sysadmin.
  - ##### `/generate all`:
    - Calls all `/generate` subcommands in the following order: `/generate roles`, `/generate categories`, `/generate channels`.
- All the creations are updated to database.
- All the commands are limited to Faculty Advisors, Coordinators and Sysadmin.

#### Files under graphql/rcos/discord_assoications and src/api/rcos/discord_assoications:
- Multiple API and .graphql files are created for "generate.rs" to update project/small groups data into database.


### Tests


Manual test:
1. Create Discord Application.
2. Create a new discord server.
3. Create roles "Faculty Advisors" and "Coordinator", add yourself to these two roles.
4. Start the docker image.
5. Invite bots to discord server using the link generated by Telescope.
6. Open Hasura console.
7. Insert Row into Semesters table: 
    - semester_id = 202201
    - title = Spring 2022
9. Insert 3 Rows(arbitrary field data) into Projects table.
10. Insert 1 Row into small group table.
    - semester_id = 202201
11. Insert 2 Rows(arbitrary field data) into users table.
12. Insert 2 Rows into enrollments table.
    - Project and user_id fields of two rows should be distinct. 
    - So we got 2 projects being active in ongoing semester and 1 project being inactive in ongoing semester.
13. Call `/generate roles` in discord server.
    - 2 project roles and 1 small group roles should be created at discord server.
    - The project_roles table and small_group_roles table should be updated.
14. Call `/generate categories` in discord server.
    - 3 categories should be created at discord server.
    - The project_categories table and small_group_categories table should be updated.
    - Newly created categories should be visible to only project/small group members, Faculty Advisors and Coordinators.
15. Call `/generate channels` in discord server.
    - 2 channels(1 voice channel and 1 text channel) should be created at each newly created category.
    - The project_channels table and small_group_channels table should be updated.
    - Newly created channels should be visible to only project/small group members, Faculty Advisors and Coordinators.
16. Delete all the discord data in database. 
17. Remove all newly project/small group categories, channels and roles from discord server.
18. Call `/generate all` to see if it produces the same result.
